pkg/controller/networkpolicy/validate_test.go

```diff
@@ -0,0 +1,952 @@
+				Spec: crdv1alpha1.ClusterNetworkPolicySpec{
+					AppliedTo: []crdv1alpha1.NetworkPolicyPeer{
+						{
+							NamespaceSelector: &metav1.LabelSelector{
+								MatchLabels: map[string]string{"foo1":"bar1"},
+							},
+						},
+					},
+					Ingress: []crdv1alpha1.Rule{
+						{
+							Action: &allowAction,
+							Ports: []crdv1alpha1.NetworkPolicyPort{
+								{
+									EndPort: &int32For80,
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedReason: "if `endPort` is specified `port` must be specified",
+		},
+		{
+			name: "acnp-endport-smaller-port",
+			policy: &crdv1alpha1.ClusterNetworkPolicy{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "acnp-endport-smaller-port",
+				},
+				Spec: crdv1alpha1.ClusterNetworkPolicySpec{
+					AppliedTo: []crdv1alpha1.NetworkPolicyPeer{
+						{
+							NamespaceSelector: &metav1.LabelSelector{
+								MatchLabels: map[string]string{"foo1":"bar1"},
+							},
+						},
+					},
+					Ingress: []crdv1alpha1.Rule{
+						{
+							Action: &allowAction,
+							Ports: []crdv1alpha1.NetworkPolicyPort{
+								{
+									Port: &int81,
+									EndPort: &int32For80,
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedReason: "`endPort` should be greater than or equal to `port`",
+		},
+		{
+			name: "acnp-named-port-with-endport",
+			policy: &crdv1alpha1.ClusterNetworkPolicy{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "acnp-named-port-with-endport",
+				},
+				Spec: crdv1alpha1.ClusterNetworkPolicySpec{
+					AppliedTo: []crdv1alpha1.NetworkPolicyPeer{
+						{
+							NamespaceSelector: &metav1.LabelSelector{
+								MatchLabels: map[string]string{"foo1":"bar1"},
+							},
+						},
+					},
+					Ingress: []crdv1alpha1.Rule{
+						{
+							Action: &allowAction,
+							Ports: []crdv1alpha1.NetworkPolicyPort{
+								{
+									Port: &strHTTP,
+									EndPort: &int32For80,
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedReason: "if `port` is a string `endPort` cannot be specified",
+		},
+		{
+			name: "acnp-port-range",
+			policy: &crdv1alpha1.ClusterNetworkPolicy{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "acnp-port-range",
+				},
+				Spec: crdv1alpha1.ClusterNetworkPolicySpec{
+					AppliedTo: []crdv1alpha1.NetworkPolicyPeer{
+						{
+							NamespaceSelector: &metav1.LabelSelector{
+								MatchLabels: map[string]string{"foo1":"bar1"},
+							},
+						},
+					},
+					Ingress: []crdv1alpha1.Rule{
+						{
+							Action: &allowAction,
+							Ports: []crdv1alpha1.NetworkPolicyPort{
+								{
+									Port: &int80,
+									EndPort: &int32For1999,
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedReason: "",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, c := newController()
+			v := NewNetworkPolicyValidator(c.NetworkPolicyController)
+			actualReason, _ := v.validateAntreaPolicy(tt.policy, nil, admv1.Create , authenticationv1.UserInfo{})
```

Comment:
curious about why we don't add a simple `assert.False` for the second return value (the boolean)?

Reply:
I was thinking we should know the reason why it fails. Because I want to test if those cases will trigger the exact failure that I expected.

Reply:
Sorry, I am not suggested replacing it, I meant adding an extra check to validate the other return value

Reply:
Oh, I misunderstood your meaning lol. Yeah, good idea. I will add it.

Reply:
BTW, in case you didn't notice, there are still some valid cases. I use "" as the reason for valid cases. 
So if we want to check the return boolean, it might need to add an additional field to `tests` making it looks like this:
```
tests := []struct {
		name           string
		policy         *crdv1alpha1.ClusterNetworkPolicy
		expectedReason string
                 expectedResult bool
	}{...}
``` 
and then use `assert.Equal(t, tt.expectedResult, actualResult)`. If this is the effort, what's your opinion?😄

Reply:
why not:
```go
if tt.expectedReason == "" {
    assert.True(t, allowed)
} else {
    assert.False(t, allowed)
}
```

Reply:
Thanks. Done.